airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/config_migrations.py

```diff
@@ -64,7 +64,85 @@ def _emit_control_message(cls, migrated_config: Mapping[str, Any]) -> None:
     def migrate(cls, args: List[str], source: SourceAmazonSellerPartner) -> None:
         """
         This method checks the input args, should the config be migrated,
-        transform if neccessary and emit the CONTROL message.
+        transform if necessary and emit the CONTROL message.
+        """
+        # get config path
+        config_path = AirbyteEntrypoint(source).extract_config(args)
+        # proceed only if `--config` arg is provided
+        if config_path:
+            # read the existing config
+            config = source.read_config(config_path)
+            # migration check
+            if cls._should_migrate(config):
+                cls._emit_control_message(cls._modify_and_save(config_path, source, config))
+
+
+class MigrateReportOptions:
+    """
+    This class stands for migrating the config at runtime,
+    while providing the backward compatibility when falling back to the previous source version.
+
+    Specifically, starting from `2.0.1`, the `account_type` property becomes required.
+    For those connector configs that do not contain this key, the default value of `Seller` will be used.
+    Reverse operation is not needed as this field is ignored in previous versions of the connector.
+    """
+
+    message_repository: MessageRepository = InMemoryMessageRepository()
+    migration_key: str = "report_options_list"
+
+    @classmethod
+    def _should_migrate(cls, config: Mapping[str, Any]) -> bool:
+        """
+        This method determines whether config requires migration.
+        Returns:
+            > True, if the transformation is necessary
+            > False, otherwise.
+        """
+        return cls.migration_key not in config and (config.get("report_options") or config.get("advanced_stream_options"))
+
+    @classmethod
+    def _transform_report_options(cls, config: Mapping[str, Any]) -> Mapping[str, Any]:
+        try:
+            report_options = json.loads(config.get("report_options", ""))
+        except json.JSONDecodeError:
+            report_options = {}
+
+        try:
+            advanced_stream_options = json.loads(config.get("advanced_stream_options", ""))
```

Comment:
Do we need to add `advance_stream_options` to `report_options`? Previously, it was used only to set attributes for the class `SellerAnalyticsSalesAndTrafficReports`.

Reply:
I'd prefer to gather this info either since `SellerAnalyticsSalesAndTrafficReports` supports `reportOptions` according to the [docs](https://developer-docs.amazon.com/sp-api/docs/report-type-values-analytics#seller-retail-analytics-reports), what do you think?

Reply:
From my understanding, `SellerAnalyticsSalesAndTrafficReports` already incorporates `reportOptions` as defined in the base class. The usage of `advanced_report_options`, though, is distinct. It was primarily for setting class attributes, perhaps to modify the window size:
```
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        if self.name in self._advanced_stream_options.keys():
            _options: dict = self._advanced_stream_options[self.name]
            if isinstance(_options, dict):
                for _option_attr, _option_val in _options.items():
                    setattr(self, _option_attr, _option_val)
```
We should consider whether to add a separate parameter for setting class attributes or to remove `advanced_stream_options` entirely, given that allowing users to set any class attribute could pose a security risk.

Reply:
Agree, deleting it, thank you!